chromium_src/chrome/browser/printing/print_preview_message_handler.cc

```diff
@@ -0,0 +1,235 @@
+}
+
+void PrintPreviewMessageHandler::OnDidGetPreviewPageCount(
+    const PrintHostMsg_DidGetPreviewPageCount_Params& params) {
+  if (params.page_count <= 0) {
+    NOTREACHED();
+    return;
+  }
+}
+
+void PrintPreviewMessageHandler::OnDidPreviewPage(
+    const PrintHostMsg_DidPreviewPage_Params& params) {
+  int page_number = params.page_number;
+  if (page_number < FIRST_PAGE_INDEX || !params.data_size)
+    return;
+}
+
+void PrintPreviewMessageHandler::OnMetafileReadyForPrinting(
+    const PrintHostMsg_DidPreviewDocument_Params& params) {
+  // Always try to stop the worker.
+  StopWorker(params.document_cookie);
+
+  if (params.expected_pages_count <= 0) {
+    NOTREACHED();
+    return;
+  }
+
+  // TODO(joth): This seems like a good match for using RefCountedStaticMemory
+  // to avoid the memory copy, but the SetPrintPreviewData call chain below
+  // needs updating to accept the RefCountedMemory* base class.
+  scoped_refptr<base::RefCountedBytes> data(
+      GetDataFromHandle(params.metafile_data_handle, params.data_size));
+  if (!data || !data->size())
+    return;
+
+  atom::NativeWindow* window = atom::NativeWindow::FromWebContents(
+      web_contents());
+  base::FilePath save_path;
+  if (!file_dialog::ShowSaveDialog(window, "Save As",
+          base::FilePath(FILE_PATH_LITERAL("print.pdf")),
+          file_dialog::Filters(), &save_path)) { // Users cancel dialog.
+    RunPrintToPDFCallback(params.preview_request_id, FAIL_CANCEL);
+    return;
+  }
+  BrowserThread::PostTaskAndReplyWithResult(
+      BrowserThread::FILE,
+      FROM_HERE,
+      base::Bind(&SavePDF, data, save_path),
+      base::Bind(&PrintPreviewMessageHandler::RunPrintToPDFCallback,
+                 base::Unretained(this),
+                 params.preview_request_id));
+}
+
+void PrintPreviewMessageHandler::OnPrintPreviewFailed(int document_cookie,
+                                                      int request_id) {
+  StopWorker(document_cookie);
+  RunPrintToPDFCallback(request_id, FAIL_PREVIEW);
+}
+
+bool PrintPreviewMessageHandler::OnMessageReceived(
+    const IPC::Message& message) {
+  bool handled = true;
+  IPC_BEGIN_MESSAGE_MAP(PrintPreviewMessageHandler, message)
+    IPC_MESSAGE_HANDLER(PrintHostMsg_DidGetPreviewPageCount,
+                        OnDidGetPreviewPageCount)
+    IPC_MESSAGE_HANDLER(PrintHostMsg_DidPreviewPage,
+                        OnDidPreviewPage)
+    IPC_MESSAGE_HANDLER(PrintHostMsg_MetafileReadyForPrinting,
+                        OnMetafileReadyForPrinting)
+    IPC_MESSAGE_HANDLER(PrintHostMsg_PrintPreviewFailed,
+                        OnPrintPreviewFailed)
+    IPC_MESSAGE_UNHANDLED(handled = false)
+  IPC_END_MESSAGE_MAP()
+  return handled;
+}
+
+void PrintPreviewMessageHandler::PrintToPDF(
+    const mate::Dictionary& options,
+    const atom::NativeWindow::PrintToPDFCallback& callback) {
+  // A simulated Chromium print preivew setting. 
+  const std::string setting_json_str =  "{ \
```

Comment:
I think we can do the setting generation in JavaScript and then pass a `base::DictionaryValue` directly here, it would make things easier when we want to add more options for printing in future.